node_modules/oae-version/lib/api.js

```diff
@@ -0,0 +1,188 @@
+var path = require('path');
+var util = require('util');
+
+var log = require('oae-logger').logger('oae-version');
+
+// A variable that will hold the path to the UI directory
+var _uiPath = null;
+
+// A variable that will hold a copy of the version information
+var _version = null;
+
+/**
+ * Initialize the version module
+ *
+ * @param  {String}     uiPath  The path to the UI directory
+ */
+var init = module.exports.init = function(uiPath) {
+    _uiPath = uiPath;
+};
+
+/**
+ * Get the version information for OAE
+ *
+ * @param  {Function}   callback                    Standard callback function
+ * @param  {Object}     callback.version            The version information
+ * @param  {String}     callback.version.hilary     The version information for Hilary
+ * @param  {String}     callback.version.3akai-ux   The version information for the UI
+ */
+var getVersion = module.exports.getVersion = function(callback) {
+    if (_version) {
+        return callback(_version);
+    }
+
+    getHilaryVersion(function(hilaryVersion) {
+        getUIVersion(function(uiVersion) {
+            _version = {
+                'hilary': hilaryVersion,
+                '3akai-ux': uiVersion
+            };
+            return callback(_version);
+        });
+    });
+};
+
+/**
+ * Get the version information for the backend
+ *
+ * @param  {Function}   callback            Standard callback function
+ * @param  {String}     callback.version    The version information for the backend
+ * @api private
+ */
+function getHilaryVersion(callback) {
+    var hilaryDirectory = path.resolve(__dirname, '..', '..', '..');
+    getVersionInfo(hilaryDirectory, callback);
+}
+
+/**
+ * Get the version information for the UI
+ *
+ * @param  {Function}   callback            Standard callback function
+ * @param  {String}     callback.version    The version information for the UI
+ * @api private
+ */
+function getUIVersion(callback) {
+    getVersionInfo(_uiPath, callback);
+}
+
+/**
+ * Get the version information for a directory. The version will be retrieved from
+ * the build info file in the root of the directory. If there's no such file present
+ * or it can't be parsed, the information will be retrieved from git. If that fails
+ * as well `Unknown` will be returned
+ *
+ * @param  {String}     directory           The directory to get the version information for
+ * @param  {Function}   callback            Standard callback function
+ * @param  {String}     callback.version    The version information for the directory
+ * @api private
+ */
+function getVersionInfo(directory, callback) {
+    getBuildVersion(directory, function(err, buildVersion) {
+        if (err) {
+            return getGitVersion(directory, function(err, gitVersion) {
+                if (err) {
+                    return callback('Unknown');
+                }
+
+                return callback(gitVersion);
+            });
+        }
+
+        return callback(buildVersion);
```

Comment:
Rather than relying on an error, we should have a function that determines if there is a build version available, then decides whether to parse a build version or use a git version instead.

Also, this is not a standard callback function as no error is ever provided as the first argument, but maybe there should be as I don't see any reason to muffle errors in this case.

If there is a build-info available, we can promote an error if it's not valid JSON, for example. It's something we'll want to be aware of and have error information for.